stubs/docker/docker/utils/build.pyi

```diff
@@ -1,38 +1,43 @@
-from _typeshed import Incomplete
+import io
+from _typeshed import Incomplete, StrOrBytesPath, StrPath
+from collections.abc import Generator, Iterable
+from os import PathLike
+from tarfile import _Fileobj
+from tempfile import _TemporaryFileWrapper
 
 def match_tag(tag: str) -> bool: ...
 def tar(
-    path,
-    exclude: Incomplete | None = None,
-    dockerfile: Incomplete | None = None,
-    fileobj: Incomplete | None = None,
+    path: PathLike[str],
+    exclude: list[str] | None = None,
+    dockerfile: tuple[str | None, str | None] | None = None,
+    fileobj: _Fileobj | None = None,
     gzip: bool = False,
-): ...
-def exclude_paths(root, patterns, dockerfile: Incomplete | None = None): ...
-def build_file_list(root): ...
+) -> _TemporaryFileWrapper[bytes] | _Fileobj: ...
+def exclude_paths(root: StrPath, patterns: list[str], dockerfile: str | None = None) -> set[str]: ...
+def build_file_list(root: str) -> list[str]: ...
 def create_archive(
-    root,
-    files: Incomplete | None = None,
-    fileobj: Incomplete | None = None,
+    root: str,
+    files: Iterable[str] | None = None,
+    fileobj: _Fileobj | None = None,
     gzip: bool = False,
     extra_files: Incomplete | None = None,
-): ...
-def mkbuildcontext(dockerfile): ...
-def split_path(p): ...
-def normalize_slashes(p): ...
-def walk(root, patterns, default: bool = True): ...
+) -> _TemporaryFileWrapper[bytes] | _Fileobj: ...
+def mkbuildcontext(dockerfile: io.IOBase | StrOrBytesPath) -> _TemporaryFileWrapper[bytes]: ...
```

Comment:
Note to future readers: Normally we try to avoid using io types directly, in favor of Protocols, but runtime explicitly checks for this class.